corehq/apps/reports/models.py

```diff
@@ -224,6 +225,46 @@ def __str__(self):
                                                       site=self.target_site)
 
 
+class TableauAppConnection(models.Model):
+    server_name = models.CharField(max_length=128)
+    site_name = models.CharField(max_length=64, default='Default')
+    app_client_id = models.CharField(max_length=64)
+    secret_id = models.CharField(max_length=64)
+    # Encrypted field. Do not set directly (use `plaintext_secret_value` instead).
+    encrypted_secret_value = models.CharField(max_length=64)
+
+    def __str__(self):
+        return '{server} {site_name} {app_client_id}'.format(server=self.server_name,
+                                                      site_name=self.site_name,
+                                                      app_client_id=self.app_client_id)
+
+    @property
+    def plaintext_secret_value(self):
+        return b64_aes_decrypt(self.encrypted_secret_value)
+
+    @plaintext_secret_value.setter
+    def plaintext_secret_value(self, plaintext):
+        self.encrypted_secret_value = b64_aes_encrypt(plaintext)
+
+
+def delete_tableau_connected_app_if_last_domain(domain):
+    try:
+        domain_details_obj = TableauDomainDetails.objects.get(domain=domain)
+        if len(TableauDomainDetails.objects.filter(app_connection=domain_details_obj.app_connection)) <= 1:
```

Comment:
Nit: `.count()` is generally preferred over `len` for querysets, as `len` will evaluate the full thing.  Also, you can use `<model_name>_set` to get the reverse of a fk relationship, like so:
```suggestion
        if domain_details_obj.app_connection.tableaudomaindetails_set.count() <= 1:
```
(Not that either of those points matter much in this instance)